internal/broker/upstream/manager.go

```diff
@@ -0,0 +1,360 @@
+	})
+}
+
+func (man *MCPManager) manage(ctx context.Context) {
+	man.logger.Debug("manage ", "upstream mcp", man.UpstreamMCP.ID())
+
+	validationStatus := man.Validate(ctx)
+	// TODO allow configuring connection attempts before removing tools
+	if !validationStatus.ConnectionStatus.IsReachable {
+		man.logger.Error("upstream is un-reachable. Removing tools from gateway", "err", validationStatus.ConnectionStatus.Error)
+		if man.removeToolsFunc != nil {
+			man.logger.Debug("tools list change remove all tools", "sever id", man.UpstreamMCP.ID())
+			man.removeTools()
+		}
+		return
+	}
+
+	if len(man.tools) == 0 {
+		man.logger.Debug("fetching tools fresh ", "serverid ", man.UpstreamMCP.ID())
+		man.setTools(ctx)
+	}
+}
+
+func (man *MCPManager) setTools(ctx context.Context) {
+	man.logger.Debug("fetching tools fresh ", "serverid ", man.UpstreamMCP.ID())
+	current, err := man.getTools(ctx)
+	if err != nil {
+		man.logger.Error("failed to get tools", "server id", man.UpstreamMCP.ID())
+		return
+	}
+	// build prefixed server tools from the fetched tools
+	newServerTools := make([]server.ServerTool, 0, len(current))
+	for _, tool := range current {
+		prefixedTool := tool
+		newServerTools = append(newServerTools, man.toolToServerTool(prefixedTool))
+	}
+
+	man.toolsLock.Lock()
+	man.tools = current
+	man.serverTools = newServerTools
+	man.toolsLock.Unlock()
+
+	if len(newServerTools) > 0 {
+		man.logger.Info("tools list change add via manage", "server id", man.UpstreamMCP.ID(), "total", len(current))
+		man.addToolsFunc(man.serverTools...)
+	}
+}
+
+// Validate checks the connection and capabilities of the upstream MCP server.
+// It attempts to connect if not already connected, pings the server, and validates
+// the protocol version and tool capabilities. Returns a ServerValidationStatus
+// containing the results.
+func (man *MCPManager) Validate(ctx context.Context) ServerValidationStatus {
+	s := ServerValidationStatus{
+		ID:                     string(man.UpstreamMCP.ID()),
+		Name:                   man.UpstreamMCP.Name,
+		ToolPrefix:             man.UpstreamMCP.ToolPrefix,
+		ConnectionStatus:       ConnectionStatus{},
+		ProtocolValidation:     ProtocolValidation{},
+		CapabilitiesValidation: CapabilitiesValidation{},
+		LastValidated:          time.Now(),
+	}
+	s.ProtocolValidation.ExpectedVersion = strings.Join(mcp.ValidProtocolVersions, ",")
+	s.ProtocolValidation.ExpectedVersion = strings.Join(mcp.ValidProtocolVersions, ",")
+	s.ConnectionStatus.IsReachable = false
+	if man.UpstreamMCP.Client == nil {
+		if err := man.UpstreamMCP.Connect(ctx); err != nil {
+			man.logger.Error("mcp manager failed to connect to server ", "server id", man.UpstreamMCP.ID(), "error", err)
+			s.ConnectionStatus.Error = err.Error()
+		}
+		s.ConnectionStatus.IsReachable = true
+	}
+	// always ping to verify connection if client exists
+	if man.UpstreamMCP.Client != nil {
```

Comment:
what happens if there was a 404 session invalid previous time and we try use the client again repeatedly?